add stale-if-error semantics

dferber90 opened the pull request:
Adds stale-if-error semantics to the Edge Config SDK.

This will reuse the previous response whenever the response to an Edge Config read has status code 500, 502, 503 or 504. This list of codes to apply stale-if-error semantics is from [Cache Control on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control).It will also reuse the previous response whenever fetch throws.

If no previous response exists the behavior is unchanged and the SDK will still throw.

There is currently no max age for stale responses. 

To decide:
- whether we want to introduce a maxAge / ttl for stale-if-error responses -> Yes
- whether we want to make stale-if-error behavior configurable via options in `createClient(process.env.EDGE_CONFIG, options)` -> Yes

dferber90 committed:
add stale-if-error semantics

dferber90 committed:
add changeset

dferber90 committed:
enter canary pre release mode

dferber90 committed:
rename CachedResponseEntry

dferber90 committed:
fix comments in tests

AndyBitz commented on packages/edge-config/src/utils/fetch-with-cached-response.ts:
```diff
@@ -9,46 +11,115 @@ interface ResponseWithCachedResponse extends Response {
-  const { headers: customHeaders = new Headers(), ...customOptions } = options;
-  const authHeader = customHeaders.get('Authorization');
-  const cacheKey = `${url},${authHeader || ''}`;
+  try {
```
I think this try-catch block should be removed.

It logs the error and then throws it, I think we only want to throw it and pass it on, to not pollute the output even if the error will be handled.

dferber90 replied:
🤦 Good catch! I added this for debugging purposes and forgot to remove it

dferber90 committed:
add DOMException polyfill for tests

dferber90 committed:
make staleIfError configurable

dferber90 commented on packages/edge-config/src/index.ts:
```diff
@@ -127,6 +127,22 @@ function getConnection(connectionString: string): Connection | null {
+   *
+   * The time is supplied in seconds. Defaults to `Infinity`.
+   */
+  staleIfError: number;
```
I made `staleIfError` configurable and pass it down to `fetchWithCachedResponse`, but I wonder if we should instead pass it using a request header directive. This would work perfectly for this case, but might not work for other kinds of options. I'm still going back and forth, but leaning towards passing it using a request header instead.

dferber90 committed:
use stale-if-error request header

transporting the information through the request header to make our fetch behave closer to an actual fetch

dferber90 committed:
add age header

dferber90 committed:
exit prerelease mode

dferber90 committed:
use short sha as snapshot
